test/nodejs/test_opal_builder.rb

```diff
@@ -4,9 +4,13 @@
 
 class TestNodejsOpalBuilder < Test::Unit::TestCase
 
+=begin
+  # throws an error: "FrozenError: can't modify frozen String"
```

Comment:
The following test is failing when strict mode is enabled.
Here's the complete stacktrace:

```
TestNodejsOpalBuilder#test_should_build_simple_ruby_file:
FrozenError: can't modify frozen String
    FrozenError: can't modify frozen String
      from <internal:corelib/runtime.js>:2446:7:in `Opal.set_encoding'
      from <internal:corelib/string/encoding.rb>:281:1:in `force_encoding'
      from parser/source/buffer.rb:92:48:in `force_encoding'
      from parser/source/buffer.rb:164:17:in `reencode_string'
      from <internal:corelib/runtime.js>:1729:5:in `Opal.send2'
      from <internal:corelib/runtime.js>:1719:5:in `Opal.send'
      from opal/compiler.rb:208:7:in `source='
      from opal/compiler.rb:198:7:in `parse'
      from opal/builder_processors.rb:84:11:in `compile'
      from opal/builder_processors.rb:0:1:in `compiled'
```

Yet another encoding issue.
Apparently, the source encoding is `ASCII-8BIT`. The parser is then calling `force_encoding` (parser/source/buffer.rb:92) to "reencode" the source to UTF-8 but "FrozenError: can't modify frozen String" is thrown!

https://github.com/opal/opal/blob/7d234115f99d9329d4f09ae9f58dd15f43ac26bf/opal/corelib/runtime.js#L2446

Reply:
Found a solution 🎉 
My first analysis was wrong!

When the buffer source is assigned, `parser` will do:

```rb
input = input.dup if input.frozen?
input = self.class.reencode_string(input)
```

https://github.com/whitequark/parser/blob/21684f12eb73ac416fa88ecee7198739a1847596/lib/parser/source/buffer.rb#L163-L164

Unfortunately, `frozen?` will return false despite the fact that `input` is a `string` primitive.
I did implement `frozen?` on String to return `true` when "self" is primitive. Not 100% sure this is a good idea?

In addition, I've added a condition in `force_encoding` to "dup" the string when we are working with a primitive (otherwise `set_encoding` will throw a FrozenError: "can't modify frozen String").


@elia Alternatively, we could call `dup` on `Opal.set_encoding` instead of throwing a `FrozenError` but it might hide issues? We could also call "dup" only if we are running in strict mode. I found this solution to check if we are effectively running in strict mode: https://stackoverflow.com/questions/10480108/is-there-any-way-to-check-if-strict-mode-is-enforced. What do you think?

Reply:
About the dup I'm a bit concerned with performance, let's treat strict mode like having `frozen_string_literal: true`, that should do the job for most things.

Since it's solved can you remove the comments?

Reply:
Done!